packages/eas-cli/src/project/publish.ts

```diff
@@ -0,0 +1,308 @@
+  );
+
+  const bundleBuffers: { [key: string]: RawAsset } = {};
+  Platforms.forEach(platform => {
+    bundleBuffers[platform] = {
+      type: 'bundle',
+      contentType: 'application/javascript',
+      buffer: fs.readFileSync(bundlePaths[platform]),
+    };
+  });
+  return bundleBuffers;
+}
+
+export function collectAssets(inputDir: string): CollectedAssets {
+  const distRoot = getDistRoot(inputDir);
+  const assets = collectUserDefinedAssets(distRoot);
+  const bundles = collectBundles(distRoot);
+  const assetsFinal: CollectedAssets = {};
+
+  let platform: PublishPlatforms;
+  for (platform of Platforms) {
+    assetsFinal[platform] = {
+      launchAsset: bundles[platform],
+      assets,
+    };
+  }
+
+  return assetsFinal;
+}
+
+export async function filterOutAssetsThatAlreadyExistAsync(
+  uniqueAssetsWithStorageKey: (RawAsset & { storageKey: string })[]
+): Promise<(RawAsset & { storageKey: string })[]> {
+  const assetMetadata = await PublishQuery.getAssetMetadataAsync(
+    uniqueAssetsWithStorageKey.map(asset => asset.storageKey)
+  );
+  const missingAssetMetadata = assetMetadata
+    .filter((am: any) => am.status !== 'EXISTS')
+    .map((am: any) => am.storageKey);
+
+  const missingAssets = uniqueAssetsWithStorageKey.filter(asset => {
+    return missingAssetMetadata.includes(asset.storageKey);
+  });
+  return missingAssets;
+}
+
+export async function uploadAssetsAsync(assetsForUpdateInfoGroup: CollectedAssets): Promise<void> {
+  let assets: RawAsset[] = [];
+  let platform: keyof CollectedAssets;
+  for (platform in assetsForUpdateInfoGroup) {
+    assets = [
+      ...assets,
+      assetsForUpdateInfoGroup[platform]!.launchAsset,
+      ...assetsForUpdateInfoGroup[platform]!.assets,
+    ];
+  }
+
+  const uniqueAssets = uniqBy<
+    RawAsset & {
+      storageKey: string;
+    }
+  >(
+    assets.map(asset => {
+      return {
+        ...asset,
+        storageKey: getStorageKeyForAsset(asset),
+      };
+    }),
+    'storageKey'
+  );
+
+  let missingAssets = await filterOutAssetsThatAlreadyExistAsync(uniqueAssets);
+  const { specifications } = await PublishMutation.getUploadURLsAsync(
+    missingAssets.map((ma: any) => ma.contentType)
+  );
+
+  await Promise.all(
+    missingAssets.map((missingAsset, i) => {
+      const presignedPost: PresignedPost = JSON.parse(specifications[i]);
+      return uploadWithPresignedPostAsync(missingAsset.buffer, presignedPost);
+    })
+  );
+
+  // Wait up to TIMEOUT_LIMIT for assets to be uploaded and processed
+  const start = Date.now();
+  while (missingAssets.length > 0) {
+    if (process.env.NODE_ENV !== 'test') {
+      await new Promise(resolve => setTimeout(resolve, 1000));
+    }
+    missingAssets = await filterOutAssetsThatAlreadyExistAsync(missingAssets);
+
+    if (Date.now() - start > TIMEOUT_LIMIT) {
+      throw new Error('Failed to upload all assets. Please try again.');
```

Comment:
Maybe add something here that makes it clearer that uploading *timed out* (rather than some other kind of error).